### ISSUE TITLE
Race conditions in initial package request can lead to wrong result sets

### ISSUE THREAD

**Issue by StephanErb** [devpi/devpi#631]
In certain situations, Devpi is returning the wrong result if a package is requested for the first time.

## Setup
* Given a pristine Devpi server that has not seen request for package `foo`.
* If two clients request `foo` at roughly the same time, Devpi will report an error to the second request. Specifically, it will report that `foo` does not exist on the server. The first request however will work as expected. All subsequent request will work as expected as well. 
* We are running an internal setup where we have to spin up new Devpi servers frequently. We are therefore seeing the issue very often. In an installation with a long-running Devpi server it is unlikely to be observed in the wild.

## Logging output
```
2019-01-09 09:54:07,137 INFO  [req172] GET /platform/Debian_9/+simple/jinja2/
2019-01-09 09:54:07,148 INFO  [req172] [Wtx70] setting projects cache for 'jinja2'
2019-01-09 09:54:07,150 INFO  [req173] GET /platform/Debian_9/+simple/jinja2/
2019-01-09 09:54:07,150 INFO  [req174] GET /platform/Debian_9/+simple/dask/
2019-01-09 09:54:07,151 INFO  [req175] GET /platform/Debian_9/+simple/psutil/
2019-01-09 09:54:07,160 INFO  [req172] [Wtx70] fswriter71: committed: keys: 'platform_mir/Debian_9/+f/3e9/d772761ec4672/Jinja2-2.6-py2-none-any.whl','platform_mir/Debian_9/jinja2/.simple','platform_mir/Debian_9/+f/399/7cf273f142420/Jinja2-2.8.1-py2.py3-none-any.whl','platform_mir/Debian_9/+f/93b/97a811b575998/Jinja2-2.9.4-py2.py3-none-any.whl','platform_mir/Debian_9/+f/f82/52e585f704269/Jinja2-2.7.3-py3-none-any.whl','platform_mir/Debian_9/+f/1cc/03ef32b64be19/Jinja2-2.8-py2.py3-none-any.whl','platform_mir/Debian_9/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl','platform_mir/Debian_9/+f/a7b/7438120dbe76a/Jinja2-2.9.5-py2.py3-none-any.whl','platform_mir/Debian_9/+f/9fe/76f4b78d1cf5d/Jinja2-2.7.3-py2-none-any.whl','platform_mir/Debian_9/+f/223/1bace0dfd8d2b/Jinja2-2.9.6-py2.py3-none-any.whl','platform_mir/Debian_9/.projects','platform_mir/Debian_9/+f/7fa/0a90a73305657/Jinja2-2.6-py3-none-any.whl'
2019-01-09 09:54:07,169 WARNI [req173] [Rtx70] Failed to check mirror whitelist. Assume it does not exists (cached not found for project jinja2)
2019-01-09 09:54:07,171 ERROR [req173] [Rtx70] while handling http://t-platform-node1.unstable.os.rack.zone:31737/platform/Debian_9/+simple/jinja2/:
The project jinja2 does not exist.
2019-01-09 09:54:07,186 INFO  [req176] GET /platform_mir/Debian_9/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl
2019-01-09 09:54:07,191 INFO  [req177] GET /platform_mir/Debian_9/+f/91c/3be4fd94a3e44/psutil-5.4.7-cp27-cp27mu-linux_x86_64.whl
2019-01-09 09:54:07,194 INFO  [req178] GET /platform_mir/Debian_9/+f/8a2/c151d5862627c/dask-1.0.0-py2.py3-none-any.whl
2019-01-09 09:54:07,198 INFO  [req176] [Wtx71] reading remote: https://software.blue-yonder.org/OSSWhitelist/generic/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl, target platform_mir/Debian_9/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl
2019-01-09 09:54:07,211 INFO  [Wtx71] fswriter72: committed: keys: 'platform_mir/Debian_9/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl', files_commit: +files/platform_mir/Debian_9/+f/74c/935a1b8bb9a39/Jinja2-2.10-py2.py3-none-any.whl
```

Please note how `req172` and `req173` both request `jinja`. Request `req172` completes as expected but `req173` fails and assumes the package does not exist. It thus returns an empty result set, leading to the aborting of the respective `pip` client.

## Potential cause (guess)
The race condition could be caused by https://github.com/devpi/devpi/blob/d98d45736fc6d1fab9c1fe5a9de175bbcd3d7b3b/server/devpi_server/extpypi.py#L459-L462 which leads to a partial information leaking out from one transaction over to the next.

## Workaround
Run the Devpi server with `--threads=1` to prevent any form of concurrency. The downside is severely reduced throughput.

**Comment by fschulze** [comment]
Your guess seems to be reasonable. I think a solution would be to set the cache value after successful transaction commit. We need a framework for that anyway to allow calling event handlers only upon a successful request/commit for several use cases where the current hooks have problems in certain cases.

**Comment by fschulze** [comment]
This is included in devpi-server 6.1.0.
